File: src/main/java/no/skatteetaten/aurora/databasehotel/dao/DatabaseHotelDataDao.java
package no.skatteetaten.aurora.databasehotel.dao;

import java.util.List;
import java.util.Map;
import java.util.Optional;

import no.skatteetaten.aurora.databasehotel.dao.dto.ExternalSchema;
import no.skatteetaten.aurora.databasehotel.dao.dto.Label;
import no.skatteetaten.aurora.databasehotel.dao.dto.SchemaData;
import no.skatteetaten.aurora.databasehotel.dao.dto.SchemaUser;

public interface DatabaseHotelDataDao {

    String SCHEMA_TYPE_MANAGED = "MANAGED";

    String SCHEMA_TYPE_EXTERNAL = "EXTERNAL";

    SchemaData createSchemaData(String name);

    SchemaData createSchemaData(String name, String schemaType);

    Optional<SchemaData> findSchemaDataById(String id);

    Optional<SchemaData> findSchemaDataByName(String name);

    void deleteSchemaData(String id);

    void deactivateSchemaData(String id);

    List<SchemaData> findAllManagedSchemaData();

    List<SchemaData> findAllSchemaDataBySchemaType(String schemaType);

    List<SchemaData> findAllManagedSchemaDataByLabels(Map<String, String> labels);

    SchemaUser createUser(String schemaId, String userType, String username, String password);

    Optional<SchemaUser> findUserById(String id);

    List<SchemaUser> findAllUsers();

    List<SchemaUser> findAllUsersForSchema(String schemaId);

    void deleteUsersForSchema(String schemaId);

    void updateUserPassword(String schemaId, String password);

    List<Label> findAllLabels();

    List<Label> findAllLabelsForSchema(String schemaId);

    void replaceLabels(String schemaId, Map<String, String> labels);

    void deleteLabelsForSchema(String schemaId);

    ExternalSchema registerExternalSchema(String id, String jdbcUrl);

    Optional<ExternalSchema> findExternalSchemaById(String id);

    void deleteExternalSchema(String schemaId);

    void updateExternalSchema(String schemaId, String username, String jdbcUrl, String password);
}

File: src/main/kotlin/no/skatteetaten/aurora/databasehotel/dao/DatabaseHotelDataDao.kt
package no.skatteetaten.aurora.databasehotel.dao

import java.time.Duration
import java.util.Date
import no.skatteetaten.aurora.databasehotel.dao.dto.ExternalSchema
import no.skatteetaten.aurora.databasehotel.dao.dto.Label
import no.skatteetaten.aurora.databasehotel.dao.dto.SchemaData
import no.skatteetaten.aurora.databasehotel.dao.dto.SchemaUser

interface DatabaseHotelDataDao {

    fun createSchemaData(name: String): SchemaData

    fun createSchemaData(name: String, schemaType: String): SchemaData

    fun findSchemaDataById(id: String, active: Boolean = true): SchemaData?

    fun findSchemaDataByName(name: String): SchemaData?

    fun findSchemaDataByNameIgnoreActive(name: String): SchemaData?

    fun deleteSchemaData(id: String)

    fun deactivateSchemaData(id: String, cooldownDuration: Duration)

    fun findAllManagedSchemaData(): List<SchemaData>

    fun findAllSchemaDataBySchemaType(schemaType: String): List<SchemaData>

    fun findAllManagedSchemaDataByLabels(labels: Map<String, String?>): List<SchemaData>

    fun findAllManagedSchemaDataIgnoreActive(): List<SchemaData>

    fun findAllManagedSchemaDataByDeleteAfterDate(deleteAfter: Date): List<SchemaData>

    fun createUser(schemaId: String, userType: String, username: String, password: String): SchemaUser

    fun findUserById(id: String): SchemaUser?

    fun findAllUsers(): List<SchemaUser>

    fun findAllUsersForSchema(schemaId: String): List<SchemaUser>

    fun deleteUsersForSchema(schemaId: String)

    fun updateUserPassword(schemaId: String, password: String)

    fun findAllLabels(): List<Label>

    fun findAllLabelsForSchema(schemaId: String): List<Label>

    fun replaceLabels(schemaId: String, labels: Map<String, String?>)

    fun deleteLabelsForSchema(schemaId: String)

    fun registerExternalSchema(id: String, jdbcUrl: String): ExternalSchema

    fun findExternalSchemaById(id: String): ExternalSchema?

    fun deleteExternalSchema(schemaId: String)

    fun updateExternalSchema(schemaId: String, username: String?, jdbcUrl: String?, password: String?)
}

object SchemaTypes {

    const val SCHEMA_TYPE_MANAGED = "MANAGED"

    const val SCHEMA_TYPE_EXTERNAL = "EXTERNAL"
}
